File: src/main/java/io/playground/orderservice/domain/saga/ProcessSaga.java
package io.playground.orderservice.domain.saga;

import lombok.AllArgsConstructor;
import lombok.Getter;

import java.time.Instant;

@Getter
@AllArgsConstructor
public class ProcessSaga {
    private Long id;

    private String orderExternalId;

    private String paymentKey;

    private ProcessSagaStatus status;

    private boolean active;

    private Instant expiresAt;

    // 재시도용 필드
    private int retryCount;

    // 재시도용 필드
    private Instant lockedUntil;

    public enum ProcessSagaStatus {
        STARTED,
        STOCKS_RESERVED,
        ORDER_CREATED,
        PAYMENT_COMPLETED,
        STOCKS_CONFIRMED,
        ORDER_COMPLETED,
        COMPENSATED
    }

    public static boolean isActiveStatus(ProcessSagaStatus status) {
        return status != ProcessSagaStatus.ORDER_COMPLETED &&
                status != ProcessSagaStatus.COMPENSATED;
    }

    public static ProcessSaga of(Long id,
                                 String orderExternalId,
                                 String paymentKey,
                                 ProcessSagaStatus status,
                                 Instant expiresAt,
                                 int retryCount,
                                 Instant lockedUntil) {
        return new ProcessSaga(
                id,
                orderExternalId,
                paymentKey,
                status,
                isActiveStatus(status),
                expiresAt,
                retryCount,
                lockedUntil
        );
    }

    public void updateStatus(ProcessSagaStatus status) {
        this.status = status;

        if (!isActiveStatus(status))
            this.active = false;
    }

    public void updateLockedUntil(Instant lockedUntil) {
        this.lockedUntil = lockedUntil;
    }
}

File: src/main/kotlin/io/playground/orderservice/domain/saga/ProcessSaga.kt
package io.playground.orderservice.domain.saga

import java.time.Instant

class ProcessSaga(
    val id: Long?,
    val orderExternalId: String,
    val paymentKey: String?,
    var status: ProcessSagaStatus,
    var active: Boolean,
    val expiresAt: Instant,
    val retryCount: Int,
    var lockedUntil: Instant?,
) {
    enum class ProcessSagaStatus {
        STARTED,
        STOCKS_RESERVED,
        ORDER_CREATED,
        PAYMENT_COMPLETED,
        STOCKS_CONFIRMED,
        ORDER_COMPLETED,
        COMPENSATED,
    }

    fun updateStatus(status: ProcessSagaStatus) {
        this.status = status
        if (!isActiveStatus(status)) {
            this.active = false
        }
    }

    fun updateLockedUntil(lockedUntil: Instant?) {
        this.lockedUntil = lockedUntil
    }

    fun isActive(): Boolean = active

    companion object {
        @JvmStatic
        fun isActiveStatus(status: ProcessSagaStatus): Boolean =
            status != ProcessSagaStatus.ORDER_COMPLETED &&
                status != ProcessSagaStatus.COMPENSATED

        @JvmStatic
        fun of(
            id: Long?,
            orderExternalId: String,
            paymentKey: String?,
            status: ProcessSagaStatus,
            expiresAt: Instant,
            retryCount: Int,
            lockedUntil: Instant?,
        ): ProcessSaga = ProcessSaga(
            id,
            orderExternalId,
            paymentKey,
            status,
            isActiveStatus(status),
            expiresAt,
            retryCount,
            lockedUntil,
        )
    }
}
